### PR TITLE
feat: Add inline loading component

### PR DESCRIPTION
**Pull request by MoizMasud**
Signed-off-by: Moiz Masud <moizkhan_007@hotmail.com>

closes https://github.com/IBM/carbon-ui-builder/issues/122

### CODE REVIEW

File: src/fragment-components/a-inline-loading.tsx [hunk 1]
```diff
@@ -0,0 +1,256 @@
...
+			imports: ['InlineLoading'],
+			code: ({ json }) => {
+				const status = `state["${nameStringToVariableString(json.codeContext?.name)}"] || "${json.status}"`;
+				return `<InlineLoading
+					onSuccess={() => {
+						if(typeof state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess === "function") {
+							return state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess;
+						}
+						return undefined;
+					})}
```
Comment (Akshat55) [inline src/fragment-components/a-inline-loading.tsx:219]:
```suggestion
					}}
```

Incorrect syntax on export for react

**Review by Akshat55 — CHANGES_REQUESTED** [review]

File: src/fragment-components/a-inline-loading.tsx [hunk 2]
```diff
@@ -0,0 +1,256 @@
...
+			value={selectedComponent.inactiveIconDescription}
+			labelText='Inactive icon description'
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				inactiveIconDescription: event.currentTarget.value
+			})} />
+		<NumberInput
+			id='successDelay'
+			min={0}
+			label='Success delay'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:106]:
```suggestion
			label='Success delay (ms)'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 3]
```diff
@@ -0,0 +1,256 @@
...
+		componentObj={componentObj}
+		headingCss={css`display: block;`}
+		rejectDrop={true}
+		{...rest}>
+			<InlineLoading
+			successDelay={componentObj.successDelay}
+			description={status[componentObj.status].description}
+			iconDescription={status[componentObj.status].iconDescription}
+			status={componentObj.status}
+			className={componentObj.cssClasses?.map((cc: any) => cc.id).join(' ')} />
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:162]:
```suggestion
				successDelay={componentObj.successDelay}
				description={status[componentObj.status].description}
				iconDescription={status[componentObj.status].iconDescription}
				status={componentObj.status}
				className={componentObj.cssClasses?.map((cc: any) => cc.id).join(' ')} />
```

File: src/fragment-components/a-inline-loading.tsx [hunk 4]
```diff
@@ -0,0 +1,256 @@
...
+		status: 'active',
+		activeText: '',
+		inactiveText: '',
+		successText: '',
+		errorText: '',
+		activeIconDescription: '',
+		errorIconDescription: '',
+		finishedIconDescription: '',
+		inactiveIconDescription: '',
+		successDelay: 1500
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:185]:
```suggestion
		status: 'active'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 5]
```diff
@@ -0,0 +1,247 @@
...
+		status: 'active'
+	},
+	image,
+	codeExport: {
+		angular: {
+			inputs: ({ json }) => `@Input() ${nameStringToVariableString(json.codeContext?.name)}Status = "${json.status}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}LoadingText = "${json.activeText}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}SuccessText = "${json.successText}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}ErrorText = "${json.errorText}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}SuccessDelay = ${json.successDelay};`,
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:185]:
```suggestion
				@Input() ${nameStringToVariableString(json.codeContext?.name)}LoadingText = "${json.activeText || 'Loading...'}";
				@Input() ${nameStringToVariableString(json.codeContext?.name)}SuccessText = "${json.successText || 'Finished.'}";
				@Input() ${nameStringToVariableString(json.codeContext?.name)}ErrorText = "${json.errorText || 'Error!'}";
				@Input() ${nameStringToVariableString(json.codeContext?.name)}SuccessDelay = ${json.successDelay === undefined ? 1500 : json.successDelay };`,
```

File: src/fragment-components/a-inline-loading.tsx [hunk 6]
```diff
@@ -0,0 +1,247 @@
...
+		react: {
+			imports: ['InlineLoading'],
+			code: ({ json }) => {
+				const status = `state["${nameStringToVariableString(json.codeContext?.name)}"] || "${json.status}"`;
+				return `<InlineLoading
+					onSuccess={() => {
+						if(typeof state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess === "function") {
+							return state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess;
+						}
+						return undefined;
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:209]:
```suggestion
							state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess();
						}
```

File: src/fragment-components/a-inline-loading.tsx [hunk 7]
```diff
@@ -0,0 +1,246 @@
...
+				</ibm-inline-loading>`;
+			}
+		},
+		react: {
+			imports: ['InlineLoading'],
+			code: ({ json }) => {
+				const status = `state["${nameStringToVariableString(json.codeContext?.name)}"] || "${json.status}"`;
+				return `<InlineLoading
+					onSuccess={() => {
+						if(typeof state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess === "function") {
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:206]:
```suggestion
						if (typeof state.${nameStringToVariableString(json.codeContext?.name)}OnSuccess === "function") {
```

File: src/fragment-components/a-inline-loading.tsx [hunk 8]
```diff
@@ -0,0 +1,246 @@
...
+					iconDescription={${nameStringToVariableString(json.codeContext?.name)}StatusIconDescription}
+					status={${status}}
+					${reactClassNamesFromComponentObj(json)} />`;
+			},
+			additionalCode: (json) => {
+				const name = nameStringToVariableString(json.codeContext?.name);
+				const status = `${name}Status`;
+				const statusDescription = `${name}StatusDescription`;
+				const statusIconDescription = `${name}StatusIconDescription`;
+				return {
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:221]:
```suggestion

				return {
```

File: src/fragment-components/a-inline-loading.tsx [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+				const name = nameStringToVariableString(json.codeContext?.name);
+				const status = `${name}Status`;
+				const statusDescription = `${name}StatusDescription`;
+				const statusIconDescription = `${name}StatusIconDescription`;
+
+				return {
+					[status]: `const ${status} = {
+						active: {
+							iconDescription: "${json.activeIconDescription}",
+							description: "${json.activeText}"
```
Comment (Akshat55) [inline src/fragment-components/a-inline-loading.tsx:226]:
```suggestion
							iconDescription: "${json.activeIconDescription || 'Loading...'}",
							description: "${json.activeText  || 'Loading...'}"
```

File: src/fragment-components/a-inline-loading.tsx [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+
+				return {
+					[status]: `const ${status} = {
+						active: {
+							iconDescription: "${json.activeIconDescription || 'Loading...'}",
+							description: "${json.activeText  || 'Loading...'}"
+						},
+						error: {
+							iconDescription: "${json.errorIconDescription}",
+							description: "${json.errorText}"
```
Comment (Akshat55) [inline src/fragment-components/a-inline-loading.tsx:230]:
```suggestion
							iconDescription: "${json.errorIconDescription || 'Error!'}",
							description: "${json.errorText || 'Error!'}"
```

File: src/fragment-components/a-inline-loading.tsx [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+						},
+						inactive: {
+							iconDescription: "${json.inactiveIconDescription}",
+							description: "${json.inactiveText}"
```
Comment (Akshat55) [inline src/fragment-components/a-inline-loading.tsx:234]:
```suggestion
							iconDescription: "${json.inactiveIconDescription || ''}",
							description: "${json.inactiveText || ''}"
```

File: src/fragment-components/a-inline-loading.tsx [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+							iconDescription: "${json.errorIconDescription || 'Error!'}",
+							description: "${json.errorText || 'Error!'}"
+						},
+						inactive: {
+							iconDescription: "${json.inactiveIconDescription || ''}",
+							description: "${json.inactiveText || ''}"
+						},
+						finished: {
+							iconDescription: "${json.finishedIconDescription}",
+							description: "${json.successText}"
```
Comment (Akshat55) [inline src/fragment-components/a-inline-loading.tsx:238]:
```suggestion
							iconDescription: "${json.finishedIconDescription || 'Finished.'}",
							description: "${json.successText || 'Finished.'}"
```

File: src/fragment-components/a-inline-loading.tsx [hunk 13]
```diff
@@ -0,0 +1,247 @@
...
+		</AComponent>
+	);
+};
+
+export const componentInfo: ComponentInfo = {
+	component: AInlineLoading,
+	codeUI: AInlineLoadingCodeUI,
+	settingsUI: AInlineLoadingSettingsUI,
+	keywords: ['inline', 'loading'],
+	name: 'Inline Loading',
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:172]:
```suggestion
	name: 'Inline loading',
```

File: src/fragment-components/a-inline-loading.tsx [hunk 14]
```diff
@@ -0,0 +1,247 @@
...
+	type: 'inline-loading',
+	defaultComponentObj: {
+		type: 'inline-loading',
+		status: 'active'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:176]:
```suggestion
		status: 'active',
		activeText: 'Loading...',
		successText: 'Finished.',
		errorText: 'Error!'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 15]
```diff
@@ -0,0 +1,247 @@
...
+	},
+	image,
+	codeExport: {
+		angular: {
+			inputs: ({ json }) => `@Input() ${nameStringToVariableString(json.codeContext?.name)}Status = "${json.status}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}LoadingText = "${json.activeText || 'Loading...'}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}SuccessText = "${json.successText || 'Finished.'}";
+				@Input() ${nameStringToVariableString(json.codeContext?.name)}ErrorText = "${json.errorText || 'Error!'}";
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:184]:
```suggestion
				@Input() ${nameStringToVariableString(json.codeContext?.name)}LoadingText = "${json.activeText}";
				@Input() ${nameStringToVariableString(json.codeContext?.name)}SuccessText = "${json.successText}";
				@Input() ${nameStringToVariableString(json.codeContext?.name)}ErrorText = "${json.errorText}";
```

File: src/fragment-components/a-inline-loading.tsx [hunk 16]
```diff
@@ -0,0 +1,247 @@
...
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				inactiveIconDescription: event.currentTarget.value
+			})} />
+		<NumberInput
+			id='successDelay'
+			min={0}
+			label='Success delay (ms)'
+			name='successDelay'
+			value={selectedComponent.successDelay}
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:108]:
```suggestion
			value={selectedComponent.successDelay}
			step={100}
```

File: src/fragment-components/a-inline-loading.tsx [hunk 17]
```diff
@@ -0,0 +1,247 @@
...
+				labelText='Finished'
+				value='finished' />
+			<RadioButton
+				id='error'
+				labelText='Error'
+				value='error' />
+		</RadioButtonGroup>
+		<TextInput
+			value={selectedComponent.activeText}
+			labelText='Active text description'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:49]:
```suggestion
			labelText='Active text'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 18]
```diff
@@ -0,0 +1,247 @@
...
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				activeText: event.currentTarget.value
+			})} />
+		<TextInput
+			value={selectedComponent.inactiveText}
+			labelText='Inactive text description'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:56]:
```suggestion
			labelText='Inactive text'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 19]
```diff
@@ -0,0 +1,247 @@
...
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				inactiveText: event.currentTarget.value
+			})} />
+		<TextInput
+			value={selectedComponent.errorText}
+			labelText='Error text description'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:63]:
```suggestion
			labelText='Error text'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 20]
```diff
@@ -0,0 +1,247 @@
...
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				errorText: event.currentTarget.value
+			})} />
+		<TextInput
+			value={selectedComponent.successText}
+			labelText='Success text description'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:70]:
```suggestion
			labelText='Success text'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 21]
```diff
@@ -0,0 +1,247 @@
...
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				successText: event.currentTarget.value
+			})} />
+		<TextInput
+			value={selectedComponent.activeIconDescription}
+			labelText='Active icon description'
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:77]:
```suggestion
			labelText='Active icon'
```

File: src/fragment-components/a-inline-loading.tsx [hunk 22]
```diff
@@ -0,0 +1,251 @@
...
+				errorText: event.currentTarget.value
+			})} />
+		<TextInput
+			value={selectedComponent.successText}
+			labelText='Success text'
+			onChange={(event: any) => setComponent({
+				...selectedComponent,
+				successText: event.currentTarget.value
+			})} />
+		<TextInput
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:75]: Put icon descriptions in an "Accessibility" accordion so they don't take that much space on the screen.

They default to the accompanying text anyway and might not even be used.

Put it at the end of the section (after all other settings)

File: src/fragment-components/a-inline-loading.tsx [hunk 23]
```diff
@@ -0,0 +1,251 @@
...
+import {
+	nameStringToVariableString,
+	reactClassNamesFromComponentObj,
+	angularClassNamesFromComponentObj
+} from '../utils/fragment-tools';
+
+export const AInlineLoadingSettingsUI = ({ selectedComponent, setComponent }: any) => {
+	return <>
+		<legend className='bx--label'>Status</legend>
+		<RadioButtonGroup
```
Comment (zvonimirfras) [inline src/fragment-components/a-inline-loading.tsx:22]: Use dropdown instead with active as default.

Users will rarely need to change this.

**Review by zvonimirfras — CHANGES_REQUESTED** [review]

**Review by zvonimirfras — APPROVED** [review]

**Pull request merged**